### PR TITLE
Create RotaryInputAccumulator

### PR DESCRIPTION
**Pull request by clickxu**
#### WHAT
Create RotaryInputAccumulator

#### WHY
Tackle similar patterns of using rotary input to adjust setting values with some touch slop and rate limiting.
Modifier.onRotaryScrollEvent is too sensitive working with RSB.

#### HOW
Uses time and distance thresholds to accumulate rotary input.

#### Checklist :clipboard:
- [x] Add explicit visibility modifier and explicit return types for public declarations
- [x] Run spotless check
- [x] Run tests
- [x] Update metalava's signature text files

### CODE REVIEW

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeScreen.kt [hunk 1]
```diff
@@ -86,7 +79,7 @@ public fun VolumeScreen(
         onAudioOutputClick = { volumeViewModel.launchOutputSelection() },
         showVolumeIndicator = showVolumeIndicator,
         focusRequester = focusRequester,
-        scrollableState = volumeViewModel.volumeScrollableState,
+        onRotaryInput = volumeViewModel::onRotaryInput,
```
Comment (yschimke) [inline audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeScreen.kt:82]: This is much nicer than the weird ScrollableState abstraction I cargo culted in.

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt [hunk 2]
```diff
@@ -42,19 +44,23 @@ import kotlinx.coroutines.flow.StateFlow
...
-    vibrator: Vibrator
+    private val vibrator: Vibrator
 ) : ViewModel() {
     public val volumeState: StateFlow<VolumeState> = volumeRepository.volumeState
 
     public val audioOutput: StateFlow<AudioOutput> = audioOutputRepository.audioOutput
 
-    public val volumeScrollableState: VolumeScrollableState =
-        VolumeScrollableState(volumeRepository, vibrator)
+    public fun onRotaryInput(change: Float) {
```
Comment (yschimke) [inline audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt:57]: Maybe deserves a comment that we assume the onRotaryInputAccumulated has coalesced changes into discrete events that signal a change is needed.  Ignoring the size of `change` iscn't clear just from this.

Perhaps

```
public fun volumeChangeByScroll(change: Int) {
```

With +1 or -1?  Someone has to make the change from Pixels to +-1, it seems strange that it's implicitly here in this method.

Unfortunately that would duplicate the logic, so I'm not sure it's worth it.  But raising for discussion.

**Review by yschimke — APPROVED** [review]
Wait for review from others, but looks good and certainly worth testing out in anger for a bit.

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt (hunk 2 above)
Comment (clickxu) [reply audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt:57]: Some thoughts on this API, if we only support +/-1 adjustment by scrolling. Could we just provide and let view to care about if it's triggered by scrolling.
```
fun increaseVolumeWithHaptics()
fun decreaseVolumeWithHaptics()
```
or even let the view to decide how many steps to adjust
```
fun adjustVolumeWithHaptics(steps: Int) {
  ...
  for (i in 1..abs(step)) // increase/decrease
  ...
}
```
If we would like to let the view model knows it's a scroll change, might be better to  pass a float as a scroll distance. WDYT?

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.horologist.compose.rotaryinput
+
+import kotlin.math.abs
+
+/** Accumulator to trigger callbacks based on rotary input event. */
+internal class RotaryInputAccumulator constructor(
```
Comment (Kpeved) [inline compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt:22]: We don't need constructor word for a primary constructor

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            eventTimeMs - lastUpdateTimeMs < rateLimitCoolDownMs
+        ) {
+            return
+        }
+        onValueChange(accumulatedDistance)
+        lastUpdateTimeMs = eventTimeMs
+        accumulatedDistance = 0f
+    }
+
+    internal companion object {
```
Comment (Kpeved) [inline compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt:60]: NIT: I'm not sure that internal word makes sense here, as the whole class is already internal

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/AccumulatedRotaryInputModifier.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import androidx.compose.ui.input.rotary.RotaryScrollEvent
+import androidx.compose.ui.input.rotary.onRotaryScrollEvent
+
+/**
+ * Accumulates the scroll distances from [RotaryScrollEvent] and notifies changes with
+ * [onValueChange] once accumulated value is over the thresholds.
+ */
+@OptIn(ExperimentalComposeUiApi::class)
+public fun Modifier.onRotaryInputAccumulated(
+    onValueChange: (change: Float) -> Unit
```
Comment (Kpeved) [inline compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/AccumulatedRotaryInputModifier.kt:30]: What do you think about providing some customisation for this modifier? Like specifying time thresholds, steps with which we'll operate? 
As an option having a similar api to Stepper or InlineSlider. Like a range and a steps count, or a step value.

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt [hunk 6]
```diff
@@ -42,19 +44,23 @@ import kotlinx.coroutines.flow.StateFlow
...
+        when {
+            change > 0 -> increaseVolume()
+            change < 0 -> decreaseVolume()
+        }
+        performHaptics()
```
Comment (Kpeved) [inline audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt:62]: Perform haptic will be triggered every time the change is received - even if we went outside of threshold ( like volume is 0 and we keep decreasing it)

**Review by Kpeved — COMMENTED** [review]
Looks good. 
Just some small comments and a question about an api

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt (hunk 2 above)
Comment (yschimke) [reply audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt:57]: I think I'm leaning towards increaseVolumeWithHaptics, sounds like you like that also?

cc @Kpeved

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/AccumulatedRotaryInputModifier.kt (hunk 5 above)
Comment (yschimke) [reply compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/AccumulatedRotaryInputModifier.kt:30]: How about as a follow up, I think that deserves it's own debate.  But I like the idea.

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt (hunk 6 above)
Comment (yschimke) [reply audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt:62]: Nice Catch - could be a follow up as I think existing code did this also.

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt (hunk 2 above)
Comment (clickxu) [reply audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt]: Let me propose the increaseVolumeWithHaptics approach.

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/GenericMotionRotaryInputAccumulator.kt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import android.view.ViewConfiguration
+import androidx.core.view.ViewConfigurationCompat
+
+/**
+ * Accumulator to trigger callbacks based on [MotionEvent] from rotary input.
+ *
+ * This should normally be passed events received from [android.view.View.onGenericMotionEvent].
+ */
+@TargetApi(Build.VERSION_CODES.R)
+public class GenericMotionRotaryInputAccumulator constructor(
```
[the remark this answers is not in the document]
Comment (clickxu) [reply compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/GenericMotionRotaryInputAccumulator.kt]: Done.

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeScreen.kt [hunk 8]
```diff
@@ -104,19 +97,12 @@ public fun VolumeScreen(
...
-            if (scrollableState != null) {
-                val coroutineScope = rememberCoroutineScope()
-
-                onRotaryScrollEvent {
-                    coroutineScope.launch {
-                        scrollableState.scrollBy(it.verticalScrollPixels)
-                    }
-                    true
-                }
+            if (onRotaryInput != null) {
```
[the remark this answers is not in the document]
Comment (clickxu) [reply audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeScreen.kt]: Done.

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/GenericMotionRotaryInputAccumulator.kt [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            event.eventTime
+        )
+        return true
+    }
+
+    /**
+     * Reads [android.view.MotionEvent] to determine whether they are rotary input events and to
+     * obtain axis value.
+     */
+    private class RotaryInputEventReader constructor(context: Context) {
```
[the remark this answers is not in the document]
Comment (clickxu) [reply compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/GenericMotionRotaryInputAccumulator.kt]: Done.

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/GenericMotionRotaryInputAccumulator.kt [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        private var scaledScrollFactor =
```
[the remark this answers is not in the document]
Comment (clickxu) [reply compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/GenericMotionRotaryInputAccumulator.kt]: Done.

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt (hunk 3 above)
Comment (clickxu) [reply compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt]: Done.

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt (hunk 4 above)
Comment (clickxu) [reply compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/RotaryInputAccumulator.kt]: Done.

File: compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/AccumulatedRotaryInputModifier.kt (hunk 5 above)
Comment (clickxu) [reply compose-layout/src/main/java/com/google/android/horologist/compose/rotaryinput/AccumulatedRotaryInputModifier.kt:30]: Agree that we have a  follow up to Integrate this logic with a Stepper.

File: audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt (hunk 6 above)
Comment (clickxu) [reply audio-ui/src/main/java/com/google/android/horologist/audio/ui/VolumeViewModel.kt:64]: A follow-up to have a similar api to Stepper or InlineSlider could help to handle this case.

**Pull request merged**